DecouplingAsymPowderMag Fitting function for Muons

sf1919 opened the pull request:
**Description of work.**

Added two new fitting functions for Decoupling of asymmetry in the ordered state of a powdered magnet fitting:-

- DecoupAsymPowderMagLong for longitudinal polarization
- DecoupAsymPowderMagRot for rotational asymmetry

Both functions use the same underlying formula and so a helper file has been added to provide this functionality to both fitting functions.

Includes tests for the helper and fitting functions and revelvant documentation.

**To test:**
1. Download [red_data.csv](https://github.com/mantidproject/mantid/files/7676093/red_data.csv)
2. Open Mantid
3. Load the red_data.csv file
4. Double click the workspace that loads and a plot should appear
5. Click Fit
6. Add a FlatBackground fitting function 
7. Add a DecoupAsymPowderMagLong fitting function and set the CharField at 1196 and fix it.
8. Run the fit

Fixes #32585 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Added source and header files

Added the source and header files for this function and updated the CMake lists

re #32585

sf1919 committed:
Created MuonHelpers files

Created source and header files for MuonHelpers and updated CMakeLists

re #32585

sf1919 committed:
Convert function to DecoupAsymPowderMagLong

Updated source and header files and CMakeLists

re #32585

sf1919 committed:
Added new rot version of function

Added header and source files for rot version of function and updated CMakeLists

re #32585

sf1919 committed:
Added documentation

Added documentation for the two new fitting functions and release notes

re #32585

sf1919 committed:
Added tests

Added tests for both new functions and the helper file. Updated Helper header and source files to deal with library linking issue.

re #32585

sf1919 committed:
Fixed formatting errors

re #32585

sf1919 committed:
Fixed formatting errors in DecoupAsymPowderMagLongTest

re #32585

sf1919 committed:
Changed log and fixed jacobians

Changed from log base 10 to natural log in getAz and updated MuonHelperTest
Fixed Jacobian derivatives and updated tests accordingly

re #32585

sf1919 committed:
Resolved cppcheck errors

re #32585

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/DecoupAsymPowderMagLong.cpp:
```diff
@@ -45,9 +45,11 @@ void DecoupAsymPowderMagLong::functionDeriv1D(Jacobian *out, const double *xValu
+    double diffasym = getAz(xValues[i], charField);
+    double diffcharfield = -xValues[i] / pow(charField, 2);
+    double b = xValues[i] / charField;
+    double diffb = ((pow(b, 2) - 1) * ((pow(b, 2) + 3) * log((b + 1.0) / (b - 1.0)) - (2 * b))) / (8 * pow(b, 4));
```
this could go in the helper too

AnthonyLim23 replied:
along with diffcharfield

sf1919 replied:
Added a new getDiffAz method to the helper.

AnthonyLim23 commented:
when trying to fit to the blue data, with initial values of Asym = 10, CharField= 1000 and A0 = 14, I get an error. If I then switch to numerical derivatives it works fine.

AnthonyLim23 commented on docs/source/release/v6.3.0/muon.rst:
```diff
@@ -86,11 +86,16 @@ Fitting Functions
+* Added an :ref:`Activation <func-Activation>` fitting function to MuonModelling Fit Functions.
+* Added two fitting functions for the decoupling of asymmetry in the ordered state of a powedered magnet for fitting.
+
+  *`DecoupAsymPowderMagLong <func-DecoupAsymPowderMagLong>` can be used for longitudinal polarization
```
the formatting on these two looks funny

sf1919 committed:
Changes after review and fix merge conflict

- Added new getDiffAz method to MuonHelpers for use in both functions
- Prevented nan/infinite values in the differentials
- Updated Muon release notes to deal with merge conflict

re #32585

sf1919 commented:
> when trying to fit to the blue data, with initial values of Asym = 10, CharField= 1000 and A0 = 14, I get an error. If I then switch to numerical derivatives it works fine.

This should be fixed now

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/DecoupAsymPowderMagLong.cpp:
```diff
@@ -45,10 +45,13 @@ void DecoupAsymPowderMagLong::functionDeriv1D(Jacobian *out, const double *xValu
-    double diffAz = diffcharfield * diffb;
+    double diffAz = getDiffAz(xValues[i], charField);
+    if (!std::isfinite(diffasym)) {
+      diffasym = 0;
```
should these just be large numbers instead of 0? Since the condition is that it has become non-finite? It does have the question of making sure that the large number has the right sign (positive or negative)

AnthonyLim23 replied:
zero is correct in the limit of b->0 https://www.wolframalpha.com/input/?i=lim+x-%3E0+log%28%28x%2B1%29%2F%28x-1%29%29%2Fx%5E4

sf1919 committed:
Add tests and tidy up

Move isfinite check to MuonHelper diff method
Added isfinite check to DecoupAsymPowderMagRot
Added tests for getDiffAz

AnthonyLim23 reviewed (approved):
This works and the code looks good

sf1919 commented:
Hi @mantidproject/gatekeepers - please can this be looked at as it is blocking work on #33090 . Many thanks
